File: web/src/repositories/cloudSyncRepository.ts
import {
  collection,
  deleteDoc,
  doc,
  getDoc,
  getDocs,
  query,
  serverTimestamp,
  setDoc,
  Timestamp,
  where,
  type Firestore,
} from 'firebase/firestore';
import type { Category, Expense, SyncedPreferences } from '@/models/types';
import { patchCategoryUpdatedAt, patchExpenseUpdatedAt } from '@/services/database';
import { newCloudId } from '@/utils/syncEntityUtils';
import {
  dedupeLegacyCategoryDocs,
  dedupeLegacyExpenseDocs,
  type DedupedPull,
  type FirestorePulledDoc,
} from '@/utils/legacyFirestoreDedup';

export interface CloudCategory extends Omit<Category, 'id'> {
  updatedAt: number;
  deleted?: boolean;
}

export interface CloudExpense extends Omit<Expense, 'id'> {
  categoryCloudId: string;
  updatedAt: number;
  deleted?: boolean;
}

function categoriesRef(db: Firestore, uid: string) {
  return collection(db, 'users', uid, 'categories');
}

function expensesRef(db: Firestore, uid: string) {
  return collection(db, 'users', uid, 'expenses');
}

function categoryDoc(db: Firestore, uid: string, cloudId: string) {
  return doc(db, 'users', uid, 'categories', cloudId);
}

function expenseDoc(db: Firestore, uid: string, cloudId: string) {
  return doc(db, 'users', uid, 'expenses', cloudId);
}

function preferencesDoc(db: Firestore, uid: string) {
  return doc(db, 'users', uid, 'preferences', 'settings');
}

function firestoreMillis(value: unknown, fallback = 0): number {
  if (typeof value === 'number' && Number.isFinite(value)) return value;
  if (
    value
    && typeof value === 'object'
    && 'toMillis' in value
    && typeof (value as { toMillis: () => number }).toMillis === 'function'
  ) {
    return (value as { toMillis: () => number }).toMillis();
  }
  return fallback;
}

function resolveCloudId(docId: string, data: Record<string, unknown>): string {
  const fromData = data.cloudId;
  if (typeof fromData === 'string' && fromData.length > 0) return fromData;
  if (docId.length >= 32 && docId.includes('-')) return docId;
  if (docId.length > 0) return docId;
  return newCloudId();
}

function categoryWritePayload(category: Category, deleted = false) {
  return {
    cloudId: category.cloudId,
    name: category.name,
    iconName: category.iconName,
    colorInt: toFirestoreColorInt(category.colorInt),
    transactionType: category.transactionType,
    sortOrder: category.sortOrder,
    updatedAt: serverTimestamp(),
    deleted,
  };
}

function expenseWritePayload(
  expense: Expense,
  categoryCloudId: string,
  deleted = false,
) {
  return {
    cloudId: expense.cloudId,
    amount: expense.amount,
    dateMillis: expense.dateMillis,
    categoryId: expense.categoryId,
    categoryCloudId,
    note: expense.note ?? '',
    receiptImagePath: cloudSafeReceiptPath(expense.receiptImagePath) ?? '',
    transactionType: expense.transactionType,
    updatedAt: serverTimestamp(),
    deleted,
  };
}

/** Signed 32-bit ARGB — matches Android Room/Firestore color ints. */
function toFirestoreColorInt(value: number): number {
  const truncated = Math.trunc(value);
  if (truncated > 0x7fffffff) return truncated - 0x1_0000_0000;
  return truncated;
}

/**
 * Firestore rules only accept receipt paths that use the portable `receipt://` scheme. Any other
 * path is dropped (sent as null) so a single legacy record can't fail the whole sync.
 */
function cloudSafeReceiptPath(path: string | null | undefined): string | null {
  return path && path.startsWith('receipt://') ? path : null;
}

function parseCategory(docId: string, data: Record<string, unknown>): CloudCategory {
  return {
    cloudId: resolveCloudId(docId, data),
    name: String(data.name ?? ''),
    iconName: String(data.iconName ?? 'shopping_bag'),
    colorInt: Number(data.colorInt ?? 0xff6a9fd4),
    transactionType: data.transactionType as Category['transactionType'],
    sortOrder: Number(data.sortOrder ?? 0),
    updatedAt: firestoreMillis(data.updatedAt),
    deleted: Boolean(data.deleted),
  };
}

function parseExpense(docId: string, data: Record<string, unknown>): CloudExpense | null {
  const cloudId = resolveCloudId(docId, data);
  if (data.deleted === true) {
    return {
      cloudId,
      amount: 0,
      dateMillis: 0,
      categoryId: Number(data.categoryId ?? 0),
      categoryCloudId: String(data.categoryCloudId ?? ''),
      note: '',
      receiptImagePath: null,
      transactionType: 'expense',
      updatedAt: firestoreMillis(data.updatedAt),
      deleted: true,
    };
  }

  const legacyCategoryId = Number(data.categoryId ?? 0);
  const resolvedCategoryCloudId = typeof data.categoryCloudId === 'string' && data.categoryCloudId.length > 0
    ? data.categoryCloudId
    : legacyCategoryId > 0
      ? String(legacyCategoryId)
      : '';
  if (!resolvedCategoryCloudId) return null;

  return {
    cloudId,
    amount: Number(data.amount ?? 0),
    dateMillis: Number(data.dateMillis ?? 0),
    categoryId: legacyCategoryId,
    categoryCloudId: resolvedCategoryCloudId,
    note: String(data.note ?? ''),
    receiptImagePath: data.receiptImagePath ? String(data.receiptImagePath) : null,
    transactionType: data.transactionType as Expense['transactionType'],
    updatedAt: firestoreMillis(data.updatedAt, firestoreMillis(data.dateMillis)),
    deleted: Boolean(data.deleted),
  };
}

export const cloudSyncRepository = {
  async pushCategory(db: Firestore, uid: string, category: Category): Promise<number> {
    const ref = categoryDoc(db, uid, category.cloudId);
    await setDoc(ref, categoryWritePayload(category), { merge: true });
    const snap = await getDoc(ref);
    return firestoreMillis(snap.data()?.updatedAt);
  },

  async pushExpense(
    db: Firestore,
    uid: string,
    expense: Expense,
    categoryCloudId: string,
  ): Promise<number> {
    const ref = expenseDoc(db, uid, expense.cloudId);
    await setDoc(ref, expenseWritePayload(expense, categoryCloudId), { merge: true });
    const snap = await getDoc(ref);
    return firestoreMillis(snap.data()?.updatedAt);
  },

  async tombstoneCategory(db: Firestore, uid: string, category: Category): Promise<number> {
    const ref = categoryDoc(db, uid, category.cloudId);
    await setDoc(
      ref,
      { deleted: true, updatedAt: serverTimestamp() },
      { merge: true },
    );
    const snap = await getDoc(ref);
    return firestoreMillis(snap.data()?.updatedAt);
  },

  async tombstoneExpense(db: Firestore, uid: string, expense: Expense): Promise<number> {
    const ref = expenseDoc(db, uid, expense.cloudId);
    await setDoc(
      ref,
      { deleted: true, updatedAt: serverTimestamp() },
      { merge: true },
    );
    const snap = await getDoc(ref);
    return firestoreMillis(snap.data()?.updatedAt);
  },

  async tombstoneExpenseCloudId(db: Firestore, uid: string, cloudId: string): Promise<void> {
    const ref = expenseDoc(db, uid, cloudId);
    await setDoc(ref, { deleted: true, updatedAt: serverTimestamp() }, { merge: true });
  },

  async tombstoneCategoryCloudId(db: Firestore, uid: string, cloudId: string): Promise<void> {
    const ref = categoryDoc(db, uid, cloudId);
    await setDoc(ref, { deleted: true, updatedAt: serverTimestamp() }, { merge: true });
  },

  async pullCategories(db: Firestore, uid: string, since = 0): Promise<DedupedPull<CloudCategory>> {
    const col = categoriesRef(db, uid);
    const snap = since > 0
      ? await getDocs(query(col, where('updatedAt', '>', Timestamp.fromMillis(since))))
      : await getDocs(col);
    const docs: FirestorePulledDoc<CloudCategory>[] = snap.docs.map((d) => ({
      docId: d.id,
      record: parseCategory(d.id, d.data() as Record<string, unknown>),
    }));
    if (since > 0) {
      return { records: docs.map((item) => item.record), legacyDocIdsToDelete: [], duplicateCloudIds: [], remapCloudIds: {} };
    }
    return dedupeLegacyCategoryDocs(docs);
  },

  async pullExpenses(db: Firestore, uid: string, since = 0): Promise<DedupedPull<CloudExpense>> {
    const col = expensesRef(db, uid);
    const snap = since > 0
      ? await getDocs(query(col, where('updatedAt', '>', Timestamp.fromMillis(since))))
      : await getDocs(col);
    const docs: FirestorePulledDoc<CloudExpense>[] = snap.docs
      .map((d) => {
        const record = parseExpense(d.id, d.data() as Record<string, unknown>);
        return record ? { docId: d.id, record } : null;
      })
      .filter((item): item is FirestorePulledDoc<CloudExpense> => item != null);
    if (since > 0) {
      return { records: docs.map((item) => item.record), legacyDocIdsToDelete: [], duplicateCloudIds: [], remapCloudIds: {} };
    }
    return dedupeLegacyExpenseDocs(docs);
  },

  async deleteLegacyDocs(
    db: Firestore,
    uid: string,
    collectionName: 'categories' | 'expenses',
    docIds: string[],
  ): Promise<void> {
    if (docIds.length === 0) return;
    const col = collectionName === 'categories' ? categoriesRef(db, uid) : expensesRef(db, uid);
    await Promise.all(docIds.map((docId) => deleteDoc(doc(col, docId))));
  },

  /**
   * Tombstones content-duplicate docs (by cloudId) so every device converges on the
   * canonical copy on its next incremental pull.
   */
  async tombstoneCloudIds(
    db: Firestore,
    uid: string,
    collectionName: 'categories' | 'expenses',
    cloudIds: string[],
  ): Promise<void> {
    if (cloudIds.length === 0) return;
    const { getDoc } = await import('firebase/firestore');
    await Promise.all(
      cloudIds.map(async (cloudId) => {
        const ref = collectionName === 'categories'
          ? categoryDoc(db, uid, cloudId)
          : expenseDoc(db, uid, cloudId);
        const snap = await getDoc(ref);
        if (!snap.exists()) return;
        await setDoc(
          ref,
          { deleted: true, updatedAt: serverTimestamp() },
          { merge: true },
        );
      }),
    );
  },

  async pushPreferences(db: Firestore, uid: string, prefs: SyncedPreferences): Promise<number> {
    const ref = preferencesDoc(db, uid);
    await setDoc(
      ref,
      {
        currency: prefs.currency,
        locale: prefs.locale === 'de' ? 'de' : 'en',
        themeMode: prefs.themeMode,
        dailyReminder: prefs.dailyReminder,
        reminderHour: prefs.reminderHour,
        reminderMinute: prefs.reminderMinute,
        analyticsPeriod: prefs.analyticsPeriod,
        recordListPeriod: prefs.recordListPeriod,
        monthlyBudget: prefs.monthlyBudget,
        updatedAt: serverTimestamp(),
      },
      { merge: true },
    );
    const snap = await getDoc(ref);
    return firestoreMillis(snap.data()?.updatedAt);
  },

  async pullPreferences(db: Firestore, uid: string): Promise<SyncedPreferences | null> {
    const snap = await getDoc(preferencesDoc(db, uid));
    if (!snap.exists()) return null;
    const data = snap.data();
    return {
      currency: String(data.currency ?? 'EUR'),
      locale: (data.locale ?? 'en') as SyncedPreferences['locale'],
      themeMode: data.themeMode as SyncedPreferences['themeMode'],
      dailyReminder: Boolean(data.dailyReminder ?? true),
      reminderHour: Number(data.reminderHour ?? 19),
      reminderMinute: Number(data.reminderMinute ?? 0),
      analyticsPeriod: String(data.analyticsPeriod ?? 'this_month'),
      recordListPeriod: data.recordListPeriod === 'all_time' ? 'all_time' : 'this_month',
      monthlyBudget: data.monthlyBudget != null ? Number(data.monthlyBudget) : null,
      updatedAt: firestoreMillis(data.updatedAt),
    };
  },

  async pushAll(
    db: Firestore,
    uid: string,
    categories: Category[],
    expenses: Expense[],
    categoryCloudIdByLocalId: Map<number, string>,
  ): Promise<void> {
    await Promise.all([
      ...categories.map(async (category) => {
        const updatedAt = await this.pushCategory(db, uid, category);
        await patchCategoryUpdatedAt(category.cloudId, updatedAt);
      }),
      ...expenses.flatMap((expense) => {
        const categoryCloudId = categoryCloudIdByLocalId.get(expense.categoryId);
        return categoryCloudId
          ? [(async () => {
              const updatedAt = await this.pushExpense(db, uid, expense, categoryCloudId);
              await patchExpenseUpdatedAt(expense.cloudId, updatedAt);
            })()]
          : [];
      }),
    ]);
  },
};

File: app/src/main/java/com/aus/ausgegeben/sync/CloudSyncRepository.kt
package com.aus.ausgegeben.sync

import com.aus.ausgegeben.data.entity.Category
import com.aus.ausgegeben.data.entity.Expense
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import kotlinx.coroutines.tasks.await

data class CloudCategory(
    val category: Category,
    val deleted: Boolean = false,
)

data class CloudExpense(
    val expense: Expense,
    val deleted: Boolean = false,
)

class CloudSyncRepository(
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance(),
) {
    private fun categoriesRef(uid: String) =
        firestore.collection("users").document(uid).collection("categories")

    private fun expensesRef(uid: String) =
        firestore.collection("users").document(uid).collection("expenses")

    private fun preferencesRef(uid: String) =
        firestore.collection("users").document(uid).collection("preferences").document("settings")

    private fun now(): Long = System.currentTimeMillis()

    suspend fun pushCategory(uid: String, category: Category) {
        if (category.id == 0L) return
        val updatedAt = category.updatedAt.takeIf { it > 0L } ?: now()
        categoriesRef(uid).document(category.id.toString()).set(
            categoryPayload(category, updatedAt),
            SetOptions.merge(),
        ).await()
    }

    suspend fun pushExpense(uid: String, expense: Expense) {
        if (expense.id == 0L) return
        val updatedAt = expense.updatedAt.takeIf { it > 0L } ?: now()
        expensesRef(uid).document(expense.id.toString()).set(
            expensePayload(expense, updatedAt),
            SetOptions.merge(),
        ).await()
    }

    suspend fun tombstoneCategory(uid: String, id: Long) {
        categoriesRef(uid).document(id.toString()).set(
            mapOf("deleted" to true, "updatedAt" to now()),
            SetOptions.merge(),
        ).await()
    }

    suspend fun tombstoneExpense(uid: String, id: Long) {
        expensesRef(uid).document(id.toString()).set(
            mapOf("deleted" to true, "updatedAt" to now()),
            SetOptions.merge(),
        ).await()
    }

    suspend fun pullCategories(uid: String): List<CloudCategory> {
        val snap = categoriesRef(uid).get().await()
        return snap.documents.mapNotNull { doc ->
            val id = doc.id.toLongOrNull() ?: return@mapNotNull null
            CloudCategory(parseCategory(id, doc.data ?: emptyMap()), doc.getBoolean("deleted") == true)
        }
    }

    suspend fun pullExpenses(uid: String): List<CloudExpense> {
        val snap = expensesRef(uid).get().await()
        return snap.documents.mapNotNull { doc ->
            val id = doc.id.toLongOrNull() ?: return@mapNotNull null
            CloudExpense(parseExpense(id, doc.data ?: emptyMap()), doc.getBoolean("deleted") == true)
        }
    }

    suspend fun pushPreferences(uid: String, prefs: SyncedPreferences) {
        preferencesRef(uid).set(
            mapOf(
                "currency" to prefs.currency,
                "locale" to prefs.locale,
                "themeMode" to prefs.themeMode,
                "dailyReminder" to prefs.dailyReminder,
                "reminderHour" to prefs.reminderHour,
                "reminderMinute" to prefs.reminderMinute,
                "analyticsPeriod" to prefs.analyticsPeriod,
                "monthlyBudget" to prefs.monthlyBudget,
                "updatedAt" to prefs.updatedAt,
            ),
            SetOptions.merge(),
        ).await()
    }

    suspend fun pullPreferences(uid: String): SyncedPreferences? {
        val snap = preferencesRef(uid).get().await()
        if (!snap.exists()) return null
        val data = snap.data ?: return null
        return SyncedPreferences(
            currency = data["currency"] as? String ?: "EUR",
            locale = data["locale"] as? String ?: "en",
            themeMode = data["themeMode"] as? String ?: "system",
            dailyReminder = data["dailyReminder"] as? Boolean ?: true,
            reminderHour = (data["reminderHour"] as? Number)?.toInt() ?: 19,
            reminderMinute = (data["reminderMinute"] as? Number)?.toInt() ?: 0,
            analyticsPeriod = data["analyticsPeriod"] as? String ?: "this_month",
            monthlyBudget = (data["monthlyBudget"] as? Number)?.toDouble(),
            updatedAt = (data["updatedAt"] as? Number)?.toLong() ?: 0L,
        )
    }

    suspend fun pushAll(uid: String, categories: List<Category>, expenses: List<Expense>) {
        categories.filter { it.id != 0L }.forEach { pushCategory(uid, it) }
        expenses.filter { it.id != 0L }.forEach { pushExpense(uid, it) }
    }

    private fun categoryPayload(category: Category, updatedAt: Long, deleted: Boolean = false) = mapOf(
        "name" to category.name,
        "iconName" to category.iconName,
        "colorInt" to category.colorInt.toLong(),
        "transactionType" to category.transactionType,
        "sortOrder" to category.sortOrder,
        "updatedAt" to updatedAt,
        "deleted" to deleted,
    )

    private fun expensePayload(expense: Expense, updatedAt: Long, deleted: Boolean = false) = mapOf(
        "amount" to expense.amount,
        "dateMillis" to expense.dateMillis,
        "categoryId" to expense.categoryId,
        "note" to expense.note,
        "receiptImagePath" to expense.receiptImagePath,
        "transactionType" to expense.transactionType,
        "updatedAt" to updatedAt,
        "deleted" to deleted,
    )

    private fun parseCategory(id: Long, data: Map<String, Any?>): Category = Category(
        id = id,
        name = data["name"] as? String ?: "",
        iconName = data["iconName"] as? String ?: "shopping_bag",
        colorInt = (data["colorInt"] as? Number)?.toInt() ?: 0xff6a9fd4.toInt(),
        transactionType = data["transactionType"] as? String ?: "expense",
        sortOrder = (data["sortOrder"] as? Number)?.toInt() ?: 0,
        updatedAt = (data["updatedAt"] as? Number)?.toLong() ?: 0L,
    )

    private fun parseExpense(id: Long, data: Map<String, Any?>): Expense = Expense(
        id = id,
        amount = (data["amount"] as? Number)?.toDouble() ?: 0.0,
        dateMillis = (data["dateMillis"] as? Number)?.toLong() ?: 0L,
        categoryId = (data["categoryId"] as? Number)?.toLong() ?: 0L,
        note = data["note"] as? String ?: "",
        receiptImagePath = data["receiptImagePath"] as? String,
        transactionType = data["transactionType"] as? String ?: "expense",
        updatedAt = (data["updatedAt"] as? Number)?.toLong()
            ?: (data["dateMillis"] as? Number)?.toLong()
            ?: 0L,
    )
}
